examples/nlp/language_modeling/megatron_gpt_export.py

```diff
@@ -0,0 +1,216 @@
+from nemo.utils import logging
+
+
+class MegatronGPTExportableModel(torch.nn.Module, Exportable):
+    def __init__(self, model):
+        super().__init__()
+        self.model = model
+        self.dtype = None
+        if model.cfg['precision'] == 'bf16':
+            self.dtype = torch.bfloat16
+        elif int(model.cfg['precision']) == 32:
+            self.dtype = torch.float
+        elif int(model.cfg['precision']) == 16:
+            self.dtype = torch.float16
+        else:
+            raise ValueError(f"precision: {model.cfg['precision']} is not supported.")
+
+    def forward(self, tokens, position_ids, attention_mask):
+        assert tokens.shape == position_ids.shape
+        assert attention_mask.shape[2] == attention_mask.shape[3] == tokens.shape[1] == position_ids.shape[1]
+        with torch.autocast('cuda', dtype=self.dtype):
+            output_tensor = self.model.forward(
+                tokens=tokens.cuda(),
+                text_position_ids=position_ids.cuda(),
+                attention_mask=attention_mask.cuda(),
+                labels=None,
+            )
+
+        return output_tensor
+
+    def freeze(self):
+        for param in self.parameters():
+            param.requires_grad = False
+
+    def input_example(self, max_batch=1, max_dim=768, seq_len=6):
+        ids = [self.model.tokenizer.text_to_ids(text) for text in ["how is the weather on           Sunday"]]
+        id_tensors = [torch.unsqueeze(torch.LongTensor(id_list), dim=0) for id_list in ids]
+        masks_and_position_ids = [
+            get_ltor_masks_and_position_ids(id_tensor, self.model.tokenizer.eos_id, False, False, False)
+            for id_tensor in id_tensors
+        ]
+        for tokens, attn_mask_and_pos_ids in zip(id_tensors, masks_and_position_ids):
+            attn_mask, _, pos_ids = attn_mask_and_pos_ids
+            return tokens, pos_ids, attn_mask
+
+    def get_dynamic_axes(self):
+        dynamic_axes = {
+            'id_tensors': {0: "BS", 1: "sequence"},
+            'position_ids': {0: "BS", 1: "sequence"},
+            'attention_mask': {0: "BS", 2: "sequence", 3: "sequence"},
+        }
+        return dynamic_axes
+
+    @property
+    def input_types(self) -> Optional[Dict[str, NeuralType]]:
+        return {
+            "id_tensors": NeuralType(('B', 'T'), ChannelType()),
+            "position_ids": NeuralType(('B', 'T'), ChannelType()),
+            "attention_mask": NeuralType(('D', 'D', 'T', 'T'), ChannelType()),
+        }
+
+    @property
+    def output_types(self) -> Optional[Dict[str, NeuralType]]:
+        return {"log_probs": NeuralType(('B', 'T', 'D'), ChannelType())}
+
+    @property
+    def input_names(self) -> List[str]:
+        return ['id_tensors', 'position_ids', 'attention_mask']
+
+    @property
+    def output_names(self) -> List[str]:
+        return ['log_probs']
+
+
+@hydra_runner(config_path="conf", config_name="megatron_gpt_export")
```

Comment:
Can this be broken into 2 parts - one override for restore_from and second for actual export() - that part can actually be part of the wrapper's forward()?
I am thinking of how this should be integrated into nemo2riva.

Reply:
I've come up with a pseudo code of your request. Is this what you had in mind?


```
@hydra_runner(config_path="conf", config_name="megatron_gpt_export")
def nemo_load(cfg):
    # load model using restore_from and return


# Need to verify Hydra support for additional args 
@hydra_runner(config_path="conf", config_name="megatron_gpt_export")
def onnx_export(cfg, nemo_model):
    # initialize and export model to ONNX


if __name__ == '__main__':
    nemo_model = nemo_load()
    nemo_export(nemo_model)
```



> second for actual export() - that part can actually be part of the wrapper's forward()?

Can you elaborate on this? As per the [docs](https://docs.nvidia.com/deeplearning/nemo/user-guide/docs/en/stable/core/export.html#exportable-hooks-and-overrides), the wrapper's forward() only contains the model forward functionality and doesn't call the export function. 


Would the above change help integrate into nemo2riva? If there's some docs on nemo2riva I can refer to for additional context, that would be great!

Reply:
What I was looking for is generalization so Nemo2RIva does not need to know specifics about the model - and there are 2 parts : the model requires customized/overloaded load (not straight restore_from) and export() call which is also special. One way to address the second part would be : model would list, say, 'mgpt_wrapper' as exportable subnet: "sub mgpt_wrapper(self): return MegatronGPTExportableModel(self)". Than, all extra stuff you had to add around .export() call for that wrapper ( with te.onnx_export(True),te.fp8_autocast( enabled=True, fp8_recipe=fp8_recipe) can just go inside MegatronGPTExportableModel.forward - so I could just call regular export() on it.

Reply:
As for the first (load from .nemo) part - I am not entirely sure how to do this transparently, we used to have some hooks for NLPRestoreConnector and Megatron. Perhaps some custom code would be necessary, I will need to take another look at this.

Reply:
Hi Boris, addressed the second part of your comment. Please review.
First part of the comment is WIP